AI/src/search/mcts/selection/PNS_UCB1.java

```diff
@@ -0,0 +1,429 @@
+	 * Updates the PN-based terms of the selection strategy for all children of current.
+	 * @param current
+	 */
+	public void updateChildrenSelectionScores(final PNMCTSNode current)
+	{
+		//  This is the array that we'll re-compute
+		final double[] childrenPNSSelectionTerms = current.childrenPNSSelectionTerms();
+		final PNMCTSNodeTypes currentNodeType = current.nodeType();
+		final int numLegalMoves = current.numLegalMoves();
+		
+		switch(pnsVariant)
+        {
+            case RANK:
+            	final List<ScoredInt> sortedChildIndices = new ArrayList<ScoredInt>();
+            	
+            	if (currentNodeType == PNMCTSNodeTypes.AND_NODE)
+            	{
+            		// AND node
+            		for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: disproof number = 1.0 for unexpanded child. TODO this correct?
+							sortedChildIndices.add(new ScoredInt(i, 1.0));
+						}
+						else
+						{
+							sortedChildIndices.add(new ScoredInt(i, child.disproofNumber()));
+						}
+					}
+            	}
+            	else
+            	{
+            		// OR node
+            		for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: proof number = 1.0 for unexpanded child. TODO this correct?
+							sortedChildIndices.add(new ScoredInt(i, 1.0));
+						}
+						else
+						{
+							sortedChildIndices.add(new ScoredInt(i, child.proofNumber()));
+						}
+					}
+            	}
+
+            	Collections.sort(sortedChildIndices, ScoredInt.ASCENDING);
+            	double lastNum = Double.NaN;
+            	int lastRank = -1;
+            	
+            	for (int i = 0; i < numLegalMoves; ++i)
+            	{
+            		final double num = sortedChildIndices.get(i).score();
+            		final int childIndex = sortedChildIndices.get(i).object();
+            		final int rank;
+            		if (lastNum == num)
+            		{
+            			// There's a tie
+            			rank = lastRank;
+            		}
+            		else
+            		{
+            			// No tie
+            			lastRank = i + 1;
+            			lastNum = num;
+            			rank = lastRank;
+            		}
+            		
+            		childrenPNSSelectionTerms[childIndex] = (1.0 - (((double) rank) / numLegalMoves));
+            	}
+
+                break;
+
+            case SUM:
+
+                double sum = 0.0;
+                
+                if (currentNodeType == PNMCTSNodeTypes.AND_NODE)
+            	{
+            		// AND node
+                	for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: disproof number = 1.0 for unexpanded child. TODO this correct?
```

Comment:
Jakub please check if this is what we want for children that have not been added to tree yet